### ISSUE TITLE
Using App Engine's MemcacheService with V6

### ISSUE THREAD

**Issue by juulderuysscher** [objectify/objectify#404]
I am finding it a little bit cumbersome to implement the MemcacheService on App Engine. Isn't it possible to have a default implementation in the library which implements a MemcacheService which can be used on App Engine? Would be great if the caching part in the documentation can be expanded. E.g. an example for how to enable caching with Memcache API on App Engine.

**Comment by gpaul-idexx** [comment]
We're running on app engine standard, and we've been attempting to write a little adapter from com.googlecode.objectify.cache.MemcacheService to com.google.appengine.api.memcache.MemcacheService but haven't gotten it working. So far, we can't seem to get any values stored in the cache. Very surprising, it seemed like it would be an easy implementation. Something about the CasValue stuff is throwing us off (I think).

In our case, it's important for us the use the cloud library, so using objectify v5 is not an option. We could certainly run our own memcached instance, but that seems silly since there is already a managed memcached version available to us in GAE Std.

Thanks

**Comment by stickfigure** [comment]
Can you put the code up somewhere? I'll be happy to review it. You're right it should be really simple.

**Comment by gpaul-idexx** [comment]
Sure. https://gist.github.com/gpaul-idexx/0c1f223bd8d33c02cc706efbb74e9a05

Thanks for having a look.

**Comment by gpaul-idexx** [comment]
Fair enough, that certainly wasn't a satisfying way to do it. Here's an amended version: https://gist.github.com/gpaul-idexx/1454c91791dbb34d98905ea33888d4bc
 
I didn't attempt to implement IdentifiableValue.withValue(Object), which I don't believe we need. Also, because that seems to be the point when things get weird. Either that withValue implementation needs to produce an instance of another implementation of objectify.IdentifiableValue, or we get back into the business of attempting to implement the IdentifiableValue.IdentifiableValue interface.

This implementation works for me in simple cases, and it works for me in my "hard" case locally on the app engine emulator. But it doesn't work on my hard case when deployed to the real app engine. It doesn't error, it just always shows a read as a cache miss.

I suspect it has something to do with it being an Entity that has subclasses and some of those subclasses hold a reference to a different subclass of that same entity. Currently this is mapped as embedded entities. I'm wondering if switching it to Ref would work better.

**Comment by stickfigure** [comment]
You're right, `IdentifiableValue.withValue()` is vestigial - I've removed it. Your code looks fine.

Are you seeing an exception in the logs? I very much doubt that having embedded entities is a problem; you'd get serialization errors. There must be something else going on. Put some logging in your adapter?

**Comment by gpaul-idexx** [comment]
No solution yet, still digging, but I figured I'd share that GAE MemcacheService treats an expiration of 0 as expire immediately, so there is a mismatch between the objectify @Cache annotation and MemcacheService where I need to turn objectify's 0 expiration into a null expiration for GAE MemcacheService.

**Comment by gpaul-idexx** [comment]
I've traced this down to the implementation of MemcacheService.getIdentifiables for the case that a key is not found in the cache. The SpyMemecacheService implementation will return a Map that contains the key, and an IdentifiableValue where the value is null. Of course, this is the behavior that objectify expects. I've replicated that in my implementation so. I've updated [this gist](https://gist.github.com/gpaul-idexx/1454c91791dbb34d98905ea33888d4bc) with the last of my changes.

**Comment by nbali** [comment]
Is it working properly now then? If yes, shouldn't this be included in the repo?

**Comment by gpaul-idexx** [comment]
Yes, it's working as best I can tell. I think it would be helpful to include something like this in objectify.
